src/legacy/ui/public/ajax_stream/ajax_stream.ts

```diff
@@ -0,0 +1,165 @@
+      const payloadStart = delimIndex + delim.length;
+
+      // We've got an incomplete batch length
+      if (delimIndex < 0) {
+        return;
+      }
+
+      const rawLen = text.slice(index, delimIndex);
+      const payloadLen = parseInt(rawLen, 10);
+      const payloadEnd = payloadStart + payloadLen;
+
+      // We've got an invalid batch message (e.g. one without a numeric length: prefix)
+      if (isNaN(payloadLen)) {
+        throw new Error(`Invalid stream response length: ${rawLen}`);
+      }
+
+      // We've got an incomplete batch message
+      if (text.length < payloadEnd) {
+        return;
+      }
+
+      const payload = JSON.parse(text.slice(payloadStart, payloadEnd));
+      handler(payload);
+
+      index = payloadEnd;
+    }
+  };
+}
+
+/**
+ * Sends an AJAX request to the server, and processes the result as a
+ * streaming HTTP/1 response.
+ *
+ * @param basePath - The Kibana basepath
+ * @param defaultHeaders - The default HTTP headers to be sent with each request
+ * @param req - The XMLHttpRequest
+ * @param opts - The request options
+ * @returns A promise which resolves when the entire batch response has been processed.
+ */
+export function ajaxStream<T>(
+  basePath: string,
+  defaultHeaders: { [k: string]: string },
+  req: XMLHttpRequestLike,
+  opts: BatchOpts<T>
+) {
+  return new Promise((resolve, reject) => {
+    const { url, method, headers } = opts;
+
+    // There are several paths by which the promise may resolve or reject. We wrap this
+    // in "once" as a safeguard against cases where we attempt more than one call. (e.g.
+    // a batch handler fails, so we reject the promise, but then new data comes in for
+    // a subsequent batch item)
+    const complete = once((err: any = undefined) => (err ? reject(err) : resolve(req)));
+
+    // Begin the request
+    req.open(method || 'POST', `${basePath}/${url.replace(/^\//, '')}`);
+    req.withCredentials = true;
+
+    // Set the HTTP headers
+    Object.entries(Object.assign({}, defaultHeaders, headers)).forEach(([k, v]) =>
+      req.setRequestHeader(k, v)
+    );
+
+    const batchHandler = processBatchResponseStream(opts.onResponse);
+    const processBatch = () => {
+      try {
+        batchHandler(req.responseText);
+      } catch (err) {
+        req.abort();
+        complete(err);
+      }
+    };
+
+    req.onprogress = processBatch;
+
+    req.onreadystatechange = () => {
+      // Older browsers don't support onprogress, so we need
+      // to call this here, too. It's safe to call this multiple
+      // times even for the same progress event.
+      processBatch();
+
+      // 4 is the magic number that means the request is done
+      if (req.readyState === 4) {
+        // 0 indicates a network failure. 400+ messages are considered server errors
+        if (req.status === 0 || req.status >= 400) {
+          complete({ message: `Batch request failed with status ${req.status}` });
```

Comment:
This would be safer as:

```js
complete(new Error(`Batch request failed with status ${req.status}`));
```

That way you're always rejecting with an Error instance instead of it sometimes being an Error and sometimes just being a plain object.

Reply:
Great point.